### PR TITLE
added validRange props in calendar to control date range

### LINKED ISSUE #9134: Calendar component's mode props can't control
### Version
3.1.4

### Environment
all

### Reproduction link
[https://codesandbox.io/s/o5ly4p25n9](https://codesandbox.io/s/o5ly4p25n9)

### Steps to reproduce
I hope Calendar's view mode could be controled outside the component itself. example use a toggle button in the parent component.

### What is expected?
when i changed mode props by parent compnent, the Calendar should switch the realistic mode.

### What is actually happening?
the Calendar didn't switch it's mode. mode props only work on the initial of Component.

### PR DESCRIPTION
**Pull request by Rohanhacker**
* [x] Make sure that you propose PR to right branch: bugfix for `master`, feature for latest active branch `feature-x.x`.
* [x] Make sure that you follow antd's [code convention](https://github.com/ant-design/ant-design/wiki/Code-convention-for-antd).
* [x] Run `npm run lint` and fix those errors before submitting in order to keep consistent code style.
* [x] Rebase before creating a PR to keep commit history clear.
* [x] Add some descriptions and refer relative issues for you PR.

Extra checklist:
**elif** *isNewFeature* **:**
  * [x] Update API docs for the component.
  * [x] Update TypeScript definition for the component.
  * [x] Add unit tests for the feature.

addresses #8645

### CODE REVIEW

File: components/calendar/__tests__/index.test.js [hunk 1]
```diff
@@ -14,4 +14,17 @@ describe('Calendar', () => {
     const value = onSelect.mock.calls[0][0];
     expect(Moment.isMoment(value)).toBe(true);
   });
+  it('only Valid range should be selectable', () => {
+    const onSelect = jest.fn();
+    const validRange = () => ({
+      startDate: Moment('2018-02-02'),
+      endDate: Moment('2018-02-18'),
+    });
```
Comment (yesmeck) [inline components/calendar/__tests__/index.test.js]: Should the return of `validRange` is a array?

Comment (Rohanhacker) [reply components/calendar/__tests__/index.test.js]: @yesmeck its an object with keys as startDate and endDate
example:
function validRange(current) {
  return {startDate: "moment date", endDate: "moment date"}
}
Do you want me to make it an array instead ?

Comment (yesmeck) [reply components/calendar/__tests__/index.test.js]: Yes I think it should be a same type as `RangePicker`'s value. Another question is why it is a function?

Comment (Rohanhacker) [reply components/calendar/__tests__/index.test.js]: so that you can control valid range dynamically in a way similar to disabledDate

Comment (Rohanhacker) [reply components/calendar/__tests__/index.test.js]: will change it to an array

Comment (Rohanhacker) [reply components/calendar/__tests__/index.test.js]: I think I should make its api similar to rangepicker with both function and array type prop

Comment (Rohanhacker) [reply components/calendar/__tests__/index.test.js]: @yesmeck updated it to be similar to rangepicker's value

**Comment by yesmeck** [comment]
@afc163 This PR is under review.

**Comment by yesmeck** [comment]
@Rohanhacker You misunderstand the `validRange`, `validRange` is different from `disabledDate`, `validRange` limits which dates the user can navigate.

**Comment by afc163** [comment]
feature-3.2 branch is removed, it should be send to `feature-3.3.0` branch.

**Comment by Rohanhacker** [comment]
@yesmeck Do you mean the header should not allow navigation to months & years which are not in the provided range ?

**Comment by yesmeck** [comment]
@Rohanhacker Yes.

**Pull request closed without merging**